aosp_diff/preliminary/frameworks/base/0001-Add-AppOverrideSettings-to-hack-app-layout-behavious.patch

```diff
@@ -0,0 +1,114 @@
+The settings is set by Andnroid prop as debug.app.override_settings=
+"<appName1>=<flags1>, <appName1>=<flags1>,..."
+flags is a bitset for override features, bit0 is forceFullscreen and
+bit1 is forceLandscape. Other bits are not defined.
+
+Signed-off-by: Marc Mao <marc.mao@intel.com>
+---
+ core/java/android/app/Activity.java           | 13 +++++
+ core/java/android/app/AppOverrideSetting.java | 55 +++++++++++++++++++
+ 2 files changed, 68 insertions(+)
+ create mode 100644 core/java/android/app/AppOverrideSetting.java
+
+diff --git a/core/java/android/app/Activity.java b/core/java/android/app/Activity.java
+index 750e3d6f6fed..5d37b299fe18 100644
+--- a/core/java/android/app/Activity.java
++++ b/core/java/android/app/Activity.java
+@@ -8037,6 +8037,18 @@ public class Activity extends ContextThemeWrapper
+         performCreate(icicle, null);
+     }
+ 
++    final void applyPhoneAppOverride() {
++        AppOverrideSetting setting = AppOverrideSetting.getInstance(getApplicationInfo().processName);
++        if (setting.forceLandscape()) {
++            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
++            Log.d("PhoneAppOverride", "Force to landscape mode");
++        }
++        if (setting.forceFullscreen()) {
++            setTheme(android.R.style.Theme_NoTitleBar_Fullscreen);
++            Log.d("PhoneAppOverride", "Force to fullscreen mode");
++        }
++    }
++
+     @UnsupportedAppUsage(maxTargetSdk = Build.VERSION_CODES.R, trackingBug = 170729553)
+     final void performCreate(Bundle icicle, PersistableBundle persistentState) {
+         if (Trace.isTagEnabled(Trace.TRACE_TAG_WINDOW_MANAGER)) {
+@@ -8051,6 +8063,7 @@ public class Activity extends ContextThemeWrapper
+         mIsInMultiWindowMode = inMultiWindowMode(windowingMode);
+         mIsInPictureInPictureMode = windowingMode == WINDOWING_MODE_PINNED;
+         restoreHasCurrentPermissionRequest(icicle);
++        applyPhoneAppOverride();
+         if (persistentState != null) {
+             onCreate(icicle, persistentState);
+         } else {
+diff --git a/core/java/android/app/AppOverrideSetting.java b/core/java/android/app/AppOverrideSetting.java
+new file mode 100644
+index 000000000000..52d4998c40d0
+--- /dev/null
++++ b/core/java/android/app/AppOverrideSetting.java
+@@ -0,0 +1,55 @@
++package android.app;
++
++import android.annotation.NonNull;
++import android.os.SystemProperties;
++import android.util.Log;
++
++public class AppOverrideSetting {
++    private static String LOG_TAG = "AppOverride";
++    private static final Object LOCK = new Object();
++    private static volatile AppOverrideSetting instance = null;
++    private static int flags = 0;
++
++    public static final int FLAG_FORCE_FULLSCREEN = (1 << 0);
++    public static final int FLAG_FORCE_LANDSCAPE = (1 << 1);
++
++    private AppOverrideSetting(@NonNull String name) {
++        String builtin = ",com.leting=3,com.cennavi.swearth=1";
```

Comment:
Suggest don't hardcode workaround apps name in the code, set builtin string as empty here, instead we can set the workaround apps name by prop(debug.app.override_settings=xxx) under device config.